new ioctl in TEE driver to register shared memory from a file descriptor (SWG-186)

etienne-lms opened the pull request:
This is a proposal for implementing the ability for a TA to safely handle memory buffer arguments related to secure buffers. See SWG-186.

This proposal defines a new ioctl in the TEE linux driver `TEE_IOC_SHM_REGISTER_FD` for userland get create a TEE shared memory object from a provided dmabuf file descriptor provided from userland.

This PR is related to changes in the OP-TEE client: https://github.com/OP-TEE/optee_client/pull/59.

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -20,25 +20,42 @@
+	struct dma_buf *dmabuf;
+	struct dma_buf_attachment *attach;
+	struct sg_table *sgt;
+};
```
I think it would be better to have a struct like:

```
struct tee_shm_dmabuf_ref {
     struct tee_shm;
     struct dma_buf *dmabuf;
     struct dma_buf_attachment *attach;
     struct sg_table *sgt
};
```

and use container_of to go from a pointer to a struct tee_shm to a pointer to a struct tee_shm_dmabuf_ref.

etienne-lms replied:
ok

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -20,25 +20,42 @@
+
+		poolm->ops->free(poolm, shm);
+	} else {
+		struct dmabuf_ref *ref = (struct dmabuf_ref *)(shm + 1);
```
So without `TEE_SHM_MAPPED` this is always what we would get? It would feel a bit safer to have something more than the absence of a flag to indicate that there's a `struct dmabuf_ref` following.

etienne-lms replied:
What about a flag `TEE_EXT_DMA_BUF`, different from `TEE_SHM_DMA_BUF`, explicitly stating that the shm object relates to both an external dmabuf reference (the one the user registered) and a local dmabuf reference (the one create by tee_shm.c to provide to userland a file descriptor for our shm object).

jenswikl replied:
`TEE_SHM_EXT_DMA_BUF` sounds good.

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -190,17 +207,109 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
+					 DMA_BIDIRECTIONAL);
+	if (ref && ref->attach)
+		dma_buf_detach(ref->dmabuf, ref->attach);
+	if (ref->dmabuf)
```
`ref &&` is missing, wouldn't it be better to surround these three ifs with a `if (ref) {...`?

etienne-lms replied:
thanks

jenswikl replied:
Missed this in the update?

etienne-lms replied:
Completly missed. My apoligies.

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -221,8 +330,12 @@ void tee_shm_free(struct tee_shm *shm)
+	 *
+	 * In case the buffer was registered (not mappable), we call
+	 * tee_shm_release directly as the initial dmabuf reference must
+	 * be released before actual shm object to released.
```
By calling `tee_shm_release()` the actual shm object is released before the dmabuf reference is released.

etienne-lms replied:
The comment is wrong: we must release the original dmabuf reference only once our local shm object is released.

jenswikl replied:
I see, but why is that needed?

etienne-lms replied:
In case of an registered external dmabuf, releasing the shm object requires releasing an sgt and an attachment on the dmabuf before releasing the dmabuf itself, as done in the proposed `tee_shm_release()`.

jenswikl replied:
`shm->dmabuf` is not the same dmabuf as the one registered or am I missing something?

etienne-lms replied:
hmmm:
I stored in `shm->dmabuf` the dmabuf reference used to track our shm object.
The user registered dmabuf reference is stored in another structure (together with the sgt resources).

This allows TEE driver to handle `shm->dmabuf` as the reference for the shm (which TEE driver is interested in): it can be used to map, close, .... the shm object.

The original dmabuf reference the user registered through a fd, is kept aside and specifically handled only at shm object creation and a shm object destruction.

jenswikl replied:
That means that calling `dma_buf_put()` would be OK here.

etienne-lms replied:
yes, you are right. (i should not be surprised)
`tee_shm_free()` should call `dma_buf_put()`.... in case the buffer was registered from kernel land and needs to be freed from kernel land.

thanks, i will fix.

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -190,17 +209,108 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
 
+struct tee_shm *tee_shm_register_fd(struct tee_context *ctx, int fd)
+{
+	struct tee_shm_dmabuf_ref *ref = NULL;
```
It seems a bit pointless to NULL initialize this variable.

etienne-lms replied:
done

etienne-lms commented on include/uapi/linux/tee.h:
```diff
@@ -256,7 +285,7 @@ struct tee_ioctl_open_session_arg {
  * tee_ioctl_param
  */
-#define TEE_IOC_OPEN_SESSION	_IOR(TEE_IOC_MAGIC, TEE_IOC_BASE + 2, \
+#define TEE_IOC_OPEN_SESSION	_IOR(TEE_IOC_MAGIC, TEE_IOC_BASE + 3, \
```
update on-going restoring legacy ioc ids.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms committed:
tee: new ioctl to a register tee_shm from a dmabuf file descriptor

This change allows userland to create a tee_shm object that refers
to a dmabuf reference.

Userland provides a dmabuf file descriptor as buffer reference.
The created tee_shm object exported as a brand new dmabuf reference
used to provide a clean fd to userland. Userland shall closed this new
fd to release the tee_shm object resources. The initial dmabuf resources
are tracked independently through original dmabuf file descriptor.

Once the buffer is registered and until it is released, TEE driver
keeps a refcount on the registered dmabuf structure.

This change only support dmabuf references that relates to physically
contiguous memory buffers.

New tee_shm flag to identify tee_shm objects built from a registered
dmabuf: TEE_SHM_EXT_DMA_BUF. Such tee_shm structures are flagged both
TEE_SHM_DMA_BUF and TEE_SHM_EXT_DMA_BUF.

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>

etienne-lms commented:
squashed, tag applied.
